tests/Test.OneDriveSdk/ChunkedUploadProviderTests.cs

```diff
@@ -0,0 +1,263 @@
+            int chunkSize,
+            long currentFileSize,
+            List<string> nextExpectedRanges)
+        {
+            this.StreamSetup(true);
+            var provider = new TestChunkedUploadProvider(
+                this.uploadSession.Object,
+                this.client.Object,
+                this.uploadStream.Object,
+                chunkSize);
+            var url = "http://myurl";
+            this.uploadStream.Setup(s => s.Length).Returns(currentFileSize);
+            var session = new UploadSession
+                {
+                    UploadUrl = url,
+                    NextExpectedRanges = nextExpectedRanges
+                };
+
+            return provider.GetRangesRemainingProxy(session);
+        }
+
+        private IEnumerable<UploadChunkRequest> SetupGetUploadChunksTest(int chunkSize, long totalSize, IEnumerable<string> ranges)
+        {
+            this.uploadSession.Object.NextExpectedRanges = ranges;
+            this.uploadStream = new Mock<Stream>();
+            this.uploadStream.Setup(s => s.Length).Returns(totalSize);
+            this.StreamSetup(true);
+
+            var provider = new ChunkedUploadProvider(
+                this.uploadSession.Object,
+                this.client.Object,
+                this.uploadStream.Object,
+                chunkSize);
+
+            return provider.GetUploadChunkRequests();
+        } 
+
+        private void AssertRangesAre(IList<Tuple<long, long>> rangesExpected, IList<Tuple<long, long>> rangesReceived)
+        {
+            Assert.AreEqual(rangesExpected.Count, rangesReceived.Count, "Unexpected number of ranges remaining");
+            for (var index = 0; index < rangesExpected.Count; index++)
+            {
+                Assert.AreEqual(
+                    rangesExpected[index],
+                    rangesReceived[index],
+                    string.Format("Expected range {0}-{1}, received {2}-{3}",
+                        rangesExpected[index].Item1,
+                        rangesExpected[index].Item2,
+                        rangesReceived[index].Item1,
+                        rangesReceived[index].Item2));
+            }
+        }
+
+        private void AssertChunksAre(IEnumerable<UploadChunkRequest> expectedChunks,
+                                     IEnumerable<UploadChunkRequest> receivedChunks)
+        {
+            Assert.AreEqual(expectedChunks.Count(), receivedChunks.Count(), "Incorrect number of chunks received");
+            var receivedSet = new HashSet<Tuple<long, long, long>>();
+            foreach (var chunk in receivedChunks)
+            {
+                receivedSet.Add(new Tuple<long, long, long>(chunk.RangeBegin, chunk.RangeEnd, chunk.TotalSessionLength));
+            }
+
+            foreach (var chunk in expectedChunks)
+            {
+                Assert.IsTrue(receivedSet.Contains(new Tuple<long, long, long>(chunk.RangeBegin, chunk.RangeEnd, chunk.TotalSessionLength)),
```

Comment:
>Contains [](start = 42, length = 8)

Instead of using Contains(), use Remove() and make sure the receivedSet is empty